### ISSUE TITLE
v1.9.0-rc3 unable to remove network due to active endpoints

### ISSUE THREAD

**Issue by tommyoshaw** [docker/docker#17555]
Getting this error :  Error response from daemon: network demo has active endpoints

I created a network called demo : docker network create demo

I had 3 containers connected to the "demo" network to verify the /etc/hosts file was being updated correctly.  I have disconnected the 3 containers and deleted them.  When trying to delete the "demo" network I get the active endpoints error.  I'm fairly sure this was working in rc2.

docker network ls
NETWORK ID          NAME                DRIVER
7cb8de790cfe        none                null                
c38cdaf251ad        host                host                
768e01129e6d        bridge              bridge              
3fe7f7cd7959        demo                bridge              
c002e04e31f7        tommy               bridge      
        
docker network rm demo
Error response from daemon: network demo has active endpoints

docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES

**Comment by tommyoshaw** [comment]
Output from inspect

docker network inspect demo
[
    {
        "name": "demo",
        "id": "3fe7f7cd79595481ebdb455b9f44daa2bc877919840b2f3ca71ad2b29e21ad21",
        "scope": "local",
        "driver": "bridge",
        "ipam": {
            "driver": "default",
            "config": [
                {}
            ]
        },
        "containers": {},
        "options": {}
    }
]

**Comment by aboch** [comment]
I quickly tried with latest master but I am not able to reproduce.

If you can consistently reproduce this issue, can you please try with a freshly created network. Just want to exclude the case where a network was created with a previous rc docker version.

If you can repro, can you also post the commands list.

Thanks!

**Comment by WeiZhang555** [comment]
This error happens only once on my machine, but I failed to reproduce. It seems to happen randomly, and I think it may caused by some deep level bug.

**Comment by aboch** [comment]
@tommyoshaw @WeiZhang555 Any suspicious docker logs ?

**Comment by WeiZhang555** [comment]
I'm sorry that the docker logs are gone so I can't give you any useful information, it's my bad.
I'll try more to reproduce this, and show you details if I succeed.

**Comment by tommyoshaw** [comment]
This is reproducible and seems to be caused by the container being started in detached mode and quickly failing while at the same time being connected to the network.

The bash script used to reproduce this is below.

If you run : 

docker network create debug

docker run -d --name test1 centos:centos6.6 /bin/bash "while(true);do echo running;sleep 30;done"

docker network connect debug test1

docker network inspect debug

docker network rm debug

This works even though the container exited with : 

/bin/bash: while(true);do echo running;sleep 30;done: No such file or directory

Which is expected since there is a missing -c

When run from a script, with docker run and docker network connect in quick succession then the network cannot be removed when the script exits.  Add in a sleep command between docker run and docker network connect fixes the problem.

# Environment

docker --version
Docker version 1.9.0-rc3, build 2100b94

docker info
Containers: 0
Images: 173
Server Version: 1.9.0-rc3
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 202
 Dirperm1 Supported: false
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.13.0-66-generic
Operating System: Ubuntu 14.04.3 LTS
CPUs: 4
Total Memory: 7.679 GiB
Name: tom-laptop
ID: RRBC:VU4D:CKIR:OGCD:IJVA:SQT2:ZMZK:INRN:H3FH:WXFM:I2RV:OXTX
Username: thshaw
Registry: https://index.docker.io/v1/

uname -a
Linux tom-laptop 3.13.0-66-generic #108-Ubuntu SMP Wed Oct 7 15:20:27 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

# Reproduce :

#!/bin/bash

echo "Creating Demo Network : docker network create demo"

docker network create demo 

echo "Create 3 named containers and attach to network"

for ctnr in node1 node2 node3;
do

docker run -d --name ${ctnr} centos:centos6.6 /bin/bash "while(true);do echo running;sleep 30;done"

#sleep 5
docker network connect demo ${ctnr}
done

# Logs

INFO[1041] POST /v1.21/networks/create                  
INFO[1042] POST /v1.21/containers/create?name=node1     
INFO[1042] POST /v1.21/containers/3f48f5ef2f72295b6a2832c35da53d0a018ce91867cb470574e790cbfd8b452c/start 
INFO[1043] POST /v1.21/networks/demo3/connect           
INFO[1043] POST /v1.21/containers/create?name=node2     
INFO[1043] POST /v1.21/containers/0c4e5271e130e0589ae3880684bb4c30ec2536b167aa230f9a3f366163ea965d/start 
INFO[1044] POST /v1.21/networks/demo3/connect           
ERRO[1044] Handler for POST /v1.21/networks/demo3/connect returned error: Container 0c4e5271e130e0589ae3880684bb4c30ec2536b167aa230f9a3f366163ea965d is not running 
ERRO[1044] HTTP Error                                    err=Container 0c4e5271e130e0589ae3880684bb4c30ec2536b167aa230f9a3f366163ea965d is not running statusCode=500
INFO[1044] POST /v1.21/containers/create?name=node3     
INFO[1045] POST /v1.21/containers/b6c95654aa89d87f952e23e118ac6977b9660327a4dcc50e606d531017714671/start 
INFO[1045] POST /v1.21/networks/demo3/connect

**Comment by thaJeztah** [comment]
would you be able to test if you're able to reproduce this on 1.9-rc4 (just to be sure?) see #17000 for the release candidates

**Comment by tommyoshaw** [comment]
Seeing the same behaviour on 1.9-rc4.

**Comment by thaJeztah** [comment]
Thanks for testing, @tommyoshaw !

**Comment by mavenugo** [comment]
@tommyoshaw thanks for reporting. we will take a look.

@thaJeztah is this critical for 1.9.0 ? I think we are at a point in 1.9.0 release cycle that we should try to identify critical issues as a must fix for this release. Would you agree ?

**Comment by thaJeztah** [comment]
@mavenugo yes, agreed; talked with @tiborvass out of bound and we agree this is not *critical* for 1.9.0. I sometimes add issues/PR to the 1.9.0 milestone so that we don't loose sight and can make the decision if it's critical or can be moved to the next (1.9.1 or 1.10) milestone.

**Comment by aboch** [comment]
@tommyoshaw 
I can reproduce the issue on v1.9.0-rc3 with your script (the one with the missing `-c` which causes the container to stop right away) immediately.

But I was not able reproduce on master, after many many attempts.
If not too many, I think it is worth to look at the extra commits in master, they can give us a clue on whether the problem has been hidden or fixed.

**Comment by tommyoshaw** [comment]
I'll triple check later. I removed the node1,2,3 containers and then ran the script against rc4 and the "zombie" network was still there. Actually not sure how to get rid of the network without reinstalling Docker.

**Comment by tommyoshaw** [comment]
Tried again with v1.9.0-rc4 and is still reproducible.  The script used to reproduce is below.
If a container fails to start and is connected to the network immediately after then the network cannot be removed.

#!/bin/bash

network_name="demo"

#remove test nodes
docker rm -f node1 node2 node3

echo "Creating Demo Network : docker network create ${network_name}"

docker network create ${network_name} 

echo "Create 3 named containers and attach to network"

for ctnr in node1 node2 node3;
do

docker run -d --name ${ctnr} centos:centos6.6 /bin/bash "while(true);do echo running;sleep 30;done"
echo "${ctnr} created"

echo "Connecting to ${network_name} network"
echo "docker network connect ${network_name} ${ctnr}"

#uncomment the sleep command and the network can be removed successfully 
#sleep 4
docker network connect ${network_name} ${ctnr}

done

#remove test nodes
docker rm -f node1 node2 node3

docker network rm ${network_name}

if [ $? -eq 0 ];then
    echo "Removed Network : ${network_name}"
else
    echo "Network not removed as expected"
    docker network ls |grep ${network_name}
fi

**Comment by mavenugo** [comment]
@tommyoshaw  thanks for the info. tried a few times and I couldnt reproduce it in RC4.
But I do believe that you have reported a valid timing issue which should be able to reproduce given the right environment. Regardless of the reproduction, we will look through the code for possible racy nature of the interactions around container start & network connect. Since this is not critical for 1.9.0, we will address it for 1.9.1.

'''
vagrant@net-1:~$ ./r.sh
Error response from daemon: no such id: node1
Error response from daemon: no such id: node2
Error response from daemon: no such id: node3
Error: failed to remove containers: [node1 node2 node3]
Creating Demo Network : docker network create demo
413d3340e31fa070a0e3cab48220d5ba9b65bf68cec3dbf6f7920ee574c213bb
Create 3 named containers and attach to network
263cabd7fb14477ac6899fc8fe5ea763a0e228b0e843d078aaf26c92588ca93a
node1 created
Connecting to demo network
docker network connect demo node1
Error response from daemon: Container 263cabd7fb14477ac6899fc8fe5ea763a0e228b0e843d078aaf26c92588ca93a is not running
cd0dfe5fe3540755f4ee6210a58dca4938ee262c6d19117cde61f39fb0a4281e
node2 created
Connecting to demo network
docker network connect demo node2
Error response from daemon: Container cd0dfe5fe3540755f4ee6210a58dca4938ee262c6d19117cde61f39fb0a4281e is not running
84c0dd0833a4270c667d4c1aa194812429b31aee0420a807d9929780ea1c9f1a
node3 created
Connecting to demo network
docker network connect demo node3
Error response from daemon: Container 84c0dd0833a4270c667d4c1aa194812429b31aee0420a807d9929780ea1c9f1a is not running
node1
node2
node3
Removed Network : demo
vagrant@net-1:~$ docker version
Client:
 Version:      1.9.0-rc4
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   e6f5a3c
 Built:        Sat Oct 31 05:35:20 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.0-rc4
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   e6f5a3c
 Built:        Sat Oct 31 05:35:20 UTC 2015
 OS/Arch:      linux/amd64
'''

**Comment by mavenugo** [comment]
@tommyoshaw as a suggestion, can you modify your script to include `set -e` so that you make sure that the `demo` network is created everytime in this script rather than trying to use an existing `demo` network that might possibly be corrupted due to some known issues from the previous RCs or older master ?

**Comment by tommyoshaw** [comment]
I updated the script each time it was run to create unique network names.  These are the networks currently left over and unable to delete.

docker network ls
NETWORK ID          NAME                DRIVER
3fe7f7cd7959        demo                bridge              
c1d1cfd14170        demo6               bridge              
53b9acdb3fd8        none                null                
74b03c8d3806        bridge              bridge              
863191df6fb0        demo6               bridge              
9dbe1d98576f        demo2               bridge              
b2278d4ca182        demo3               bridge              
c002e04e31f7        tommy               bridge              
c8c2826e4bf0        demo6               bridge              
ceb5a0de1bc2        demo10              bridge              
d8e966b6e125        demo9               bridge              
ac650772f84f        host                host                
5372d70d982e        demo12              bridge              
a0e3ef5a4c43        demo7               bridge              
b8a821c4e9c6        demo5               bridge

**Comment by jsindy** [comment]
Spun up docker 1.9 and swarm 1.0 today. Came across this issue.

Created a overlay network with swarm. Spun up 4 containers connected to this network. rm -f all containers. Tried to network rm and got the same error.

Error response from daemon: 500 Internal Server Error: network taco has active endpoints

Tried reproducing with no luck.

**Comment by jsindy** [comment]
I'll update if I see it again.

**Comment by nathanleclaire** [comment]
I am experiencing the same issue. Happy to post whatever information is needed, or @mavenugo @mrjana I can let you poke at my setup.

**Comment by tiborvass** [comment]
Closed by #17558

**Comment by aboch** [comment]
Fixed by https://github.com/docker/docker/pull/17558
Sorry for the confusion, we thought the above fix made it in 1.9.0.

**Comment by nazar-pc** [comment]
I'm experiencing this with stable versions of Engine, Machine, Swarm and Compose:
```
nazar-pc@nazar-pc /m/Data> docker network rm data
Error response from daemon: 500 Internal Server Error: network data has active endpoints
nazar-pc@nazar-pc /m/Data> docker network inspect data
[
    {
        "Name": "data",
        "Id": "f0e39717b681460838b933a5e2f82b2091acd887732cb041f0e032879f93c882",
        "Scope": "global",
        "Driver": "overlay",
        "IPAM": {
            "Driver": "default",
            "Config": [
                {}
            ]
        },
        "Containers": {},
        "Options": {}
    }
]
nazar-pc@nazar-pc /m/Data> docker network ls
NETWORK ID          NAME                      DRIVER
f0e39717b681        data                      overlay             
399779f86337        slave-2/docker_gwbridge   bridge              
ba84c78de1e6        master/bridge             bridge              
006c3431c913        slave-4/docker_gwbridge   bridge              
20fa9067c15b        slave-4/none              null                
b5e9c765d325        slave-4/host              host                
df178c963d96        slave-2/host              host                
5045752ee3d8        slave-2/none              null                
a062f1ecae8e        slave-3/none              null                
21aa91a08a6f        slave-3/host              host                
e7561bd51c41        slave-4/bridge            bridge              
b4567d3eba63        slave-2/bridge            bridge              
eab9e3a07b1c        slave-3/bridge            bridge              
d10b35d8741e        slave-3/docker_gwbridge   bridge              
ddd15e316402        master/none               null                
0de5c6b92f27        master/host               host                
nazar-pc@nazar-pc /m/Data> docker info
Containers: 5
Images: 20
Role: primary
Strategy: spread
Filters: health, port, dependency, affinity, constraint
Nodes: 4
 master: 192.168.99.101:2376
  └ Containers: 2
  └ Reserved CPUs: 0 / 1
  └ Reserved Memory: 0 B / 517.6 MiB
  └ Labels: executiondriver=native-0.2, kernelversion=4.1.12-boot2docker, operatingsystem=Boot2Docker 1.9.0 (TCL 6.4); master : 16e4a2a - Tue Nov  3 19:49:22 UTC 2015, provider=virtualbox, storagedriver=aufs
 slave-2: 192.168.99.103:2376
  └ Containers: 1
  └ Reserved CPUs: 0 / 1
  └ Reserved Memory: 0 B / 517.6 MiB
  └ Labels: executiondriver=native-0.2, kernelversion=4.1.12-boot2docker, operatingsystem=Boot2Docker 1.9.0 (TCL 6.4); master : 16e4a2a - Tue Nov  3 19:49:22 UTC 2015, provider=virtualbox, storagedriver=aufs
 slave-3: 192.168.99.104:2376
  └ Containers: 1
  └ Reserved CPUs: 0 / 1
  └ Reserved Memory: 0 B / 517.6 MiB
  └ Labels: executiondriver=native-0.2, kernelversion=4.1.12-boot2docker, operatingsystem=Boot2Docker 1.9.0 (TCL 6.4); master : 16e4a2a - Tue Nov  3 19:49:22 UTC 2015, provider=virtualbox, storagedriver=aufs
 slave-4: 192.168.99.105:2376
  └ Containers: 1
  └ Reserved CPUs: 0 / 1
  └ Reserved Memory: 0 B / 517.6 MiB
  └ Labels: executiondriver=native-0.2, kernelversion=4.1.12-boot2docker, operatingsystem=Boot2Docker 1.9.0 (TCL 6.4); master : 16e4a2a - Tue Nov  3 19:49:22 UTC 2015, provider=virtualbox, storagedriver=aufs
CPUs: 4
Total Memory: 2.022 GiB
Name: 20a86d99da81
```

**Comment by salanki** [comment]
I am experiencing this issue on Docker version 1.9.1, build a34a1d5. Removing container, stopping docker  rm /var/lib/docker/network/files/local-kv.db does not work as a workaround. What more info do you need from me, and do you want a new bug opened?

**Comment by thaJeztah** [comment]
@salanki perhaps it's better to create a new issue yes, but add a link to this issue. At least provide the output of `docker version` and `docker info`. If possible, a way to reproduce, and if you have something special in your setup (also, e.g. if you're using Swarm or Docker directly).
